### PR TITLE
execv: check if file is executed

### PR DESCRIPTION
**Pull request by niewim19**
JIRA: PD-213

## Description

Added check in `execv()` function that checks if path (after resolving) leads to a regular file as it is possible to _execute a directory_ which is nonsense

## Motivation and Context

Issue from which this PR arised:
 - https://github.com/phoenix-rtos/phoenix-rtos-project/issues/248

As stated in documentation:
 > The exec family of functions shall replace the current process image with a new process image. The new image shall be constructed from a regular, executable file called the new process image file.

So i added check via `stat()` if resolved path is indeed regular file.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (ia32-generic).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: unistd/sys.c [hunk 1]
```diff
@@ -138,7 +140,13 @@ int execve(const char *file, char *const argv[], char *const envp[])
 	if ((canonical_path = resolve_path(file, NULL, 1, 0)) == NULL)
 		return -1; /* errno set by resolve_path */
 
-	err = exec(canonical_path, argv, envp);
+	/* check if it is file */
+	err = stat(canonical_path, &buf);
+	if (!err && !S_ISREG(buf.st_mode))
+		err = -ENOENT;
```
Comment (agkaminski) [inline unistd/sys.c:146]:
```suggestion
		err = -ENOEXEC;
```

Comment (agkaminski) [reply unistd/sys.c:146]: Behaviour confirmed on linux/glibc

**Review by agkaminski — APPROVED** [review]

File: unistd/sys.c [hunk 2]
```diff
@@ -60,6 +61,22 @@ static int shebang(const char *path)
 	return fd;
 }
 
+static int iself(const char *path)
```
[the remark this answers is not in the document]
Comment (niewim19) [reply unistd/sys.c:64]: obviously not an error, but I will change it so it does not produce an spelling error :)

**Comment by niewim19** [comment]
in `unistd: execve()` added check if we are executing an elf file. Further down in kernel it is also checked, however there is some kind of wrong propagation of error which leads to system hangup after trying to execute a non-ELF file. SO check here, however redundant, is necessary for quality.

Additional issue will be created

File: unistd/sys.c [hunk 3]
```diff
@@ -60,6 +61,22 @@ static int shebang(const char *path)
...
+	unsigned char elfhead[4] = { 0x7F, 'E', 'L', 'F' };
+
+	if ((fd = open(path, O_RDONLY)) < 0)
+		return -1;
+
+	if ((read(fd, head, 4) < 0) || memcmp(head, elfhead, 4)) {
+		close(fd);
+		return -1;
+	}
+	return 0;
```
Comment (kemonats) [inline unistd/sys.c:77]:
```suggestion
	if ((read(fd, head, 4) < 4) || memcmp(head, elfhead, 4)) {
		close(fd);
		return -1;
	}
	close(fd);
	return 0;
```
Missing `close(fd);`
You should check how many bytes were actually read.

Comment (niewim19) [reply unistd/sys.c]: Thank you! Added missing `close()`

File: unistd/sys.c [hunk 4]
```diff
@@ -60,6 +61,23 @@ static int shebang(const char *path)
 	return fd;
 }
 
+static int checkelf(const char *path)
```
Comment (agkaminski) [inline unistd/sys.c:64]: Wouldn't this screw up scripts executions (#!/bin/bash)?

Comment (niewim19) [reply unistd/sys.c:64]: If we execute a script this comes into play:
```c
if ((fd = shebang(file)) >= 0) {
(...)
	while (*interp == ' ') ++interp;
	end = interp;
	while (*end && !isspace(*end)) ++end;
	*end = 0;
(...)
	file = interp;
	argv = sb_args;
}
```

which substitutes the `file` path with interpreter path (for `#!/bin/bash` -> `/bin/bash`) which is later resolved to interpreter binary file which is good.

Also I checked against `psh` scripts adding a file called `abc` which contains only:
> !/bin/psh

and it produces this output:
> psh: /abc is not a psh script

which means that passing to script works well. After all, syscall `exec()` is **only** for binary files. Only arguments change so user sees it as _executing a script_

**Comment by nalajcie** [comment]
IMHO ELF verification should not be done in userspace as nothing prevents developer from using syscall directly (and this adds another time penalty for `exec()`). if we know there is a bug in error propagation in kernel, we should fix it there.

I'm not sure about whether we should verify if a file is a normal file in kernel also or it should/could be done in userspace only.

Side note: we're not checking anywhere if a file has executable bit set but I assume it's not related to this PR (mentioning it as it could easily be added now that we're calling `stat()` just before the exec).

**Comment by damianloew** [comment]
Running scripts seems to work properly with these changes. Tested using the busybox test suite.

File: unistd/sys.c [hunk 5]
```diff
@@ -60,6 +61,23 @@ static int shebang(const char *path)
...
+{
+	int fd;
+	unsigned char head[4];
+	unsigned char elfhead[4] = { 0x7F, 'E', 'L', 'F' };
+
+	if ((fd = open(path, O_RDONLY)) < 0)
+		return -1;
+
+	if ((read(fd, head, 4) < 0) || memcmp(head, elfhead, 4)) {
```
Comment (kemonats) [inline unistd/sys.c:73]:
```suggestion
	if ((read(fd, head, 4) < 4) || memcmp(head, elfhead, 4)) {
```
Why give up checking the actual number of bytes read?
Why compare against the ELF header if less than 4 bytes were read.

File: unistd/sys.c [hunk 6]
```diff
@@ -60,6 +61,23 @@ static int shebang(const char *path)
 	return fd;
 }
 
+static int checkelf(const char *path)
+{
```
Comment (agkaminski) [inline unistd/sys.c:65]: We can drop this after/if https://github.com/phoenix-rtos/phoenix-rtos-kernel/pull/269 is merged

File: unistd/sys.c (hunk 5 above)
Comment (niewim19) [reply unistd/sys.c:73]: Yes. Good notice. Unfortunately after recent commit to kernel this is handled in kernel so while `checkelf()` addition is being dropped. However thank you for an advice!

**Review by agkaminski — APPROVED** [review]

File: unistd/sys.c (hunk 6 above)
Comment (niewim19) [reply unistd/sys.c]: Mentioned commit merged. Whole `checkelf()` addition was removed from commit

**Comment by niewim19** [comment]
Removed ELF check as it is now correctly handled after https://github.com/phoenix-rtos/phoenix-rtos-kernel/pull/269

Merging without executable bit check as our native `psh` does not have any control over it (no `chmod` equivalent yet) and maybe this should be done also in kernel (so that we won`t execute restricted file via direct call to `exec()` from process with lower restrictions)

**Pull request merged**